knowledge-base/adr/0031-resilience-options.md

```diff
@@ -0,0 +1,310 @@
+|   rate limit    |    ❌     |   ❌   |      n.a.      |    n.a.     |
+|   bulk limit    |    ❌     |   ❌   |      n.a.      |    n.a.     |
+
+## Decision
+
+- Rate limit and bulk limit we will not be done -> no user request.
+- We will implement an [extendable middleware approach]('#variant-c').
+- Step 1: Circuit breaker is in resilience middleware and made tenant aware
+- Step 2: Retry is in resilience middleware and excluded if circuit breaker is open.
+- Step 3 (Optional): Make resilience globally configurable for all requests.
+  Do this on demand or after customer feedback.
+- Default for circuit breaker is on.
+- Default for retry is off since this is a bigger behavior change
+
+|     Option      | On target | On BTP | Default target | Default BTP | Remarks                                    |
+| :-------------: | :-------: | :----: | :------------: | :---------: | ------------------------------------------ |
+| circuit breaker |    ✅     |   ✅   |    enabled     |   enabled   | tenant aware                               |
+|     timeout     |    ✅     |   ✅   |    enabled     |   enabled   |                                            |
+|      retry      |    ✅     |   ✅   |    disabled    |  disabled   | no retry: circuit breaker open and 401,403 |
+|   rate limit    |    ❌     |   ❌   |      n.a.      |    n.a.     |                                            |
+|   bulk limit    |    ❌     |   ❌   |      n.a.      |    n.a.     |                                            |
+
+## Consequences
+
+The user has the option to switch-on and adjust resilience.
+If options are not sufficient a custom implementation can be used.
+
+## Implementation Details
+
+### Options
+
+- For retry, we will use [async retry](https://www.npmjs.com/package/async-retry)
+- For circuit breaker we will use [opossum](https://www.npmjs.com/package/opossum)
+
+This determines the possible options you can set.
+The `RetryOptions` and `CircuitBreakerOptions` could be used to overwrite the default values.
+If you pass `true`,this will enable the resilience option with the default values.
+If you pass `false`, this will disable the resilience option.
+
+```ts
+type RetryOptions = undefined | true | false | AsynRetryLibOptions;
+type CircuitBreakerOptions = undefined | true | false | OpssumLibOption;
+type TimeoutOptions = undefined | number | { service: number; target: number };
+
+interface OpssumLibOption {
+  timeout?: number | false | undefined; //default 10000
+  errorThresholdPercentage?: number | undefined; //default 50
+  volumeThreshold?: number | undefined; // default 10
+  resetTimeout?: number | undefined; //default 30000
+  isolationStragtegy?: IsolationStrategy; //default tenant
+}
+
+interface AsynRetryLibOptions {
+  retries?: number; //default 10
+  factor?: number; // default  2.
+  minTimeout?: number; // default 1000 ms.
+  maxTimeout?: number; // default Infinity.
+  randomize?: boolean; // default true.
```

Comment:
```suggestion
  factor?: number; // default 2
  minTimeoutInMs?: number; // default 1000 ms
  maxTimeoutInMs?: number; // default Infinity
  randomize?: boolean; // default true
```

Reply:
Maybe we could use https://github.com/vercel/ms for timeout times (it makes it clear and easy which time unit is used instead of always guessing if it is seconds / milliseconds / ...)

Reply:
I like the lib, but what will happen for JS users?

Reply:
The lib uses strings like '2min' or '10ms'. So it is the same no matter if you use TS or JS.

Reply:
sure, I had more in mind, that the user does not get a design time failure if he/she passes `1000` instead of `'1sec'` string.

Reply:
I think the `ms(value)` should throw a reasonable error if I understood it right.